From 1996a8258084fb29b2d5fc3e8bbfa0c1e334381f Mon Sep 17 00:00:00 2001
From: Richard Sandiford <rdsandiford@googlemail.com>
Date: Fri, 20 Feb 2026 11:51:12 +0000
Subject: [PATCH] rtl-ssa: Relax RPO requirement for insn_changes [PR134994]

The RTL-SSA insn_changes framework expected all changes to be
presented in reverse post-order.  While that's reasonable for
changes within the same EBB, it isn't reasonable for changes
in different EBBs, since things like phi def-use links have
no defined order. (Keeping them ordered would make updates
too expensive in the worst case.)

As shown in the PR, late-combine was already not honouring this rule
when it had to look through degenerate phi uses.  This caused
verify_insn_changes to walk off the end of the insn list, even though
the change set was actually feasible.

This patch relaxes the rule accordingly.  I can't think of a specific
situation in which the null dereference would still occur with the new
rule, if callers follow it and if they check that each change's move
range is individually valid.  Still, it's hard to prove, and Jeff's
null pointer check makes conceptual sense for this kind of list walk.
I therefore think we should do both, as a belt-and-braces fix.

gcc/
	PR rtl-optimization/123994
	* doc/rtl.texi: Do not require RTL-SSA changes from different EBBs
	to be in RPO.
	* rtl-ssa/changes.cc (function_info::change_insns): Likewise.
	(function_info::verify_insn_changes): Likewise.  Avoid running off
	the end of the insn list.

gcc/testsuite/
	PR rtl-optimization/123994
	* gcc.dg/torture/pr1123991.c: New test.

Co-authored-by: Jeff Law <jeffrey.law@oss.qualcomm.com>
---
 gcc/doc/rtl.texi                        |  6 ++--
 gcc/rtl-ssa/changes.cc                  | 35 +++++++++++++++------
 gcc/testsuite/gcc.dg/torture/pr123991.c | 41 +++++++++++++++++++++++++
 3 files changed, 69 insertions(+), 13 deletions(-)
 create mode 100644 gcc/testsuite/gcc.dg/torture/pr123991.c

diff --git a/gcc/doc/rtl.texi b/gcc/doc/rtl.texi
index d29cbcde3b8..f1de3969247 100644
--- a/gcc/doc/rtl.texi
+++ b/gcc/doc/rtl.texi
@@ -5053,9 +5053,9 @@ attempt.  It should then construct an @code{rtl_ssa::insn_change}
 for each change that it wants to make.
 
 After this, it should combine the changes into a sequence of
-@code{rtl_ssa::insn_change} pointers.  This sequence must be in
-reverse postorder; the instructions will remain strictly in the
-order that the sequence specifies.
+@code{rtl_ssa::insn_change} pointers.  All changes to the same extended
+basic block must be grouped together and be listed in the order that
+the pass wants them to appear within that extended basic block.
 
 For example, if a pass is changing exactly two instructions,
 it might do:
diff --git a/gcc/rtl-ssa/changes.cc b/gcc/rtl-ssa/changes.cc
index e0df982880c..d8b9fa8dbe9 100644
--- a/gcc/rtl-ssa/changes.cc
+++ b/gcc/rtl-ssa/changes.cc
@@ -103,16 +103,22 @@ function_info::verify_insn_changes (array_slice<insn_change *const> changes)
   CLEAR_HARD_REG_SET (defined_hard_regs);
   CLEAR_HARD_REG_SET (clobbered_hard_regs);
 
-  insn_info *min_insn = m_first_insn;
+  insn_info *min_insn = nullptr;
   for (insn_change *change : changes)
     if (!change->is_deletion ())
       {
-	// Make sure that the changes can be kept in their current order
-	// while honoring all of the move ranges.
-	min_insn = later_insn (min_insn, change->move_range.first);
-	while (min_insn != change->insn () && !can_insert_after (min_insn))
+	// Make sure that changes in the same EBB can be kept in their current
+	// order while honoring all of the move ranges.
+	auto *ebb = change->move_range.first->ebb ();
+	if (!min_insn || min_insn->ebb () != ebb)
+	  min_insn = change->move_range.first;
+	else
+	  min_insn = later_insn (min_insn, change->move_range.first);
+	while (min_insn
+	       && min_insn != change->insn ()
+	       && !can_insert_after (min_insn))
 	  min_insn = min_insn->next_nondebug_insn ();
-	if (*min_insn > *change->move_range.last)
+	if (!min_insn || *min_insn > *change->move_range.last)
 	  {
 	    if (dump_file && (dump_flags & TDF_DETAILS))
 	      fprintf (dump_file, "no viable insn position assignment\n");
@@ -755,7 +761,8 @@ function_info::change_insns (array_slice<insn_change *> changes)
 {
   auto watermark = temp_watermark ();
 
-  insn_info *min_insn = m_first_insn;
+  auto_bitmap ebbs;
+  insn_info *min_insn = nullptr;
   for (insn_change *change : changes)
     {
       // Tentatively mark all the old uses and definitions for deletion.
@@ -774,7 +781,14 @@ function_info::change_insns (array_slice<insn_change *> changes)
 	    update_notes (change->rtl ());
 
 	  // Make sure that the placement of this instruction would still
-	  // leave room for previous instructions.
+	  // leave room for previous instructions in the same EBB.
+	  auto *ebb = change->move_range.first->ebb ();
+	  if (!min_insn || min_insn->ebb () != ebb)
+	    {
+	      min_insn = ebb->phi_insn ();
+	      // Make sure that all changes to the same EBB are consecutive.
+	      gcc_assert (bitmap_set_bit (ebbs, ebb->first_bb ()->index ()));
+	    }
 	  change->move_range = move_later_than (change->move_range, min_insn);
 	  if (!canonicalize_move_range (change->move_range, change->insn ()))
 	    // verify_insn_changes is supposed to make sure that this holds.
@@ -809,8 +823,9 @@ function_info::change_insns (array_slice<insn_change *> changes)
 	delete_insn (change);
       else
 	{
-	  // Make sure that this instruction comes before later ones.
-	  if (following_insn)
+	  // Make sure that this instruction comes before later ones
+	  // in the same EBB.
+	  if (following_insn && following_insn->ebb () == change.ebb ())
 	    {
 	      change.move_range = move_earlier_than (change.move_range,
 						     following_insn);
diff --git a/gcc/testsuite/gcc.dg/torture/pr123991.c b/gcc/testsuite/gcc.dg/torture/pr123991.c
new file mode 100644
index 00000000000..0f60aba4408
--- /dev/null
+++ b/gcc/testsuite/gcc.dg/torture/pr123991.c
@@ -0,0 +1,41 @@
+/* { dg-additional-options "-w" } */
+/* { dg-additional-options "-march=rv64gcb -mabi=lp64d -misa-spec=20191213 -mtls-dialect=trad" { target riscv64*-*-* } } */
+
+typedef unsigned char uint8_t;
+typedef signed char int8_t;
+typedef unsigned long long uint64_t;
+typedef  long long int64_t;
+typedef int int32_t;
+typedef unsigned uint32_t;
+typedef short int16_t;
+typedef unsigned short uint16_t;
+
+
+#define BS_VEC(type, num) type __attribute__((vector_size(num * sizeof(type))))
+uint8_t backsmith_snippet_141(int16_t, uint64_t);
+static int32_t backsmith_snippet_122(uint64_t BS_ARG_0, uint32_t BS_ARG_1)
+{
+    BS_ARG_0 =
+        BS_ARG_1 ? (BS_VEC(uint64_t, 16)){}[BS_ARG_0] : -23;
+    return BS_ARG_0;
+}
+uint16_t backsmith_pure_0(uint64_t BS_ARG_2, uint32_t BS_ARG_3)
+{
+    int64_t BS_VAR_0[6];
+    int8_t BS_VAR_3[80];
+    for (uint16_t BS_INC_0 = 0; BS_INC_0 < 8; BS_INC_0 += 1)
+    {
+        uint64_t BS_TEMP_590 = BS_INC_0;
+        int64_t t =
+        BS_VAR_0[BS_INC_0] = BS_INC_0
+            ? (BS_TEMP_590 ? BS_ARG_2 >> BS_TEMP_590 : 0)
+                ?: backsmith_snippet_141(0, BS_INC_0)
+            : 0;
+        BS_VAR_3[BS_INC_0] =
+            backsmith_snippet_122(BS_VAR_0[ (uint64_t)BS_INC_0 ] < 0,
+                                  BS_ARG_3)
+            ?: BS_ARG_3;
+    }
+    if (BS_ARG_2) BS_VAR_3[BS_ARG_3 < 80 ? BS_ARG_3 : 0] = 0;
+    return BS_VAR_3[4];
+}
-- 
2.53.0

